### ISSUE TITLE
Mising token on logout

### ISSUE THREAD

**Issue by skodapetr** [Wikidata/Wikidata-Toolkit#428]
Exception:
```
Caused by: org.wikidata.wdtk.wikibaseapi.apierrors.TokenErrorException: [notoken] The "token" parameter must be set.
	at org.wikidata.wdtk.wikibaseapi.apierrors.MediaWikiApiErrorHandler.throwMediaWikiApiErrorException(MediaWikiApiErrorHandler.java:55)
	at org.wikidata.wdtk.wikibaseapi.ApiConnection.checkErrors(ApiConnection.java:466)
	at org.wikidata.wdtk.wikibaseapi.ApiConnection.sendJsonRequest(ApiConnection.java:406)
	at org.wikidata.wdtk.wikibaseapi.ApiConnection.logout(ApiConnection.java:339)
```
Code to reproduce:
```
        ApiConnection connection = new ApiConnection("https://www.wikidata.org/w/api.php");
        connection.login("username", "pswd");
        connection.logout();
```
The logout method sends following parameters:
```
assert -> user
format -> json
action -> logout
```
however the [documentation](https://www.wikidata.org/w/api.php?action=help&modules=logout) suggest that the token parameter is required as indicated by the exception.

ApiConnection is deprecated, it is suggested to use PasswordApiConnection instead. However, I was not able to find that class in this repository. I was able to find BasicApiConnection, but it suffers from the same issue. Is there any other class that should be used?

The same code works fine on our demo instance with version MediaWiki 1.32.0 but the wikidata.org is of version 1.34.0-wmf.17 (8268fd7) with a lot more extensions. 
Our version also does not require (according to documentation) the token parameter for logout.

**Comment by wetneb** [comment]
Hi @skodapetr, thanks for this report. Sorry about the confusion between PasswordApiConnection and BasicApiConnection, indeed it is BasicApiConnection that we meant to recommend. This deprecation is just a first step towards a migration to a setup where ApiConnection is an interface that is implemented by various login methods (basic password, OAuth, 2FA…).

It looks like it would be good to add a token indeed (if it also works with instances where the token is not required?)
Any PR in this direction would be most welcome, if you want to give it a go.

**Comment by skodapetr** [comment]
@wetneb It seems that using the token works with both instances I tested with (1.32.0, 1.34.0-wmf.17), so the change should be backward compatible.

The token for the logout action is the CSRF token. This token is not stored/used in the connection, instead, it is fetched only when needed by the WbEditingAction class. That is why it is not available in the connection class. I would like to create PR, but I do not know what is the preferred way how to make the token available. I see a few options, but there might be more.

1. Request the token before logout. This should be fine as requesting the same token multiple times, return the same token value. But seems a little odd to request something just for logout.
2. Connection class may monitor token request and store tokens. So if WbEditingAction requested the token before, the token can be reused in the logout method. But this introduces additional state beside cookies that need to be managed.

**Comment by wetneb** [comment]
Good points! I think both options would be fine, but maybe option 2 would be better on the long run. Access to the CSRF token is arguably something that ought to be held in the connection rather than the editing module, since it can be used for logout and probably other non-editing actions (adding a page to a watchlist, say).

**Comment by wetneb** [comment]
Closed by #429.
